libglusterfs/src/inode.c

```diff
@@ -765,7 +765,19 @@ inode_forget_atomic(inode_t *inode, uint64_t nlookup)
         GF_ATOMIC_INIT(inode->nlookup, 0);
     } else {
         inode_lookup = GF_ATOMIC_FETCH_SUB(inode->nlookup, nlookup);
-        GF_ASSERT(inode_lookup >= nlookup);
+        /*
+          GF_ASSERT(inode_lookup >= nlookup) assert may fail due to kernel bug
+          more details can found on links:
+          https://github.com/gluster/glusterfs/pull/4081
+          https://github.com/gluster/glusterfs/issues/4074
+        */
+        if (inode_lookup < nlookup) {
+            GF_ATOMIC_FETCH_ADD(inode->nlookup, nlookup - inode_lookup);
+            gf_msg_callingfn(THIS->name, GF_LOG_CRITICAL, 0,
+                             LG_MSG_ASSERTION_FAILED,
+                             "GF_ASSERT(inode_lookup >= nlookup) may fail due "
+                             "to kernel bug, reset inode->nlookup to 0");
```

Comment:
Do we indeed reset inode->nlookup to 0? And if we do, is the way to do it via GF_ATOMIC_FETCH_ADD() ?

Reply:
This is necessary to make sure that other concurrent updates don't happen between reading the current value and updating it.

Reply:
Then why not GF_ATOMIC_INIT() ?

Reply:
An init could overwrite a modification made from another thread.

The bug here is that the kernel could send a forget for more lookups than it has really sent. So we need to adjust the counter as if it had received these additional lookups. Though unlikely, it may happen that another thread is also modifying the counter, so just adding the difference instead of unconditionally setting it to 0 seems safer.

On the other hand, this should happen very rarely or even cease to happen once the kernel bug is fixed, so the cost of an atomic here is mostly irrelevant.